src/tools/rpgmodel.h

```diff
@@ -0,0 +1,74 @@
+#ifndef RPGMODEL_H
+#define RPGMODEL_H
+
+#include <QObject>
+#include <QAbstractItemModel>
+#include <data.h>
+
+template <class DATA>
+class RpgModel : public QAbstractListModel
+{
+public:
+    RpgModel(QObject *parent = 0) : QAbstractListModel(parent), _data(DATA()()) {}
+    virtual ~RpgModel() {}
+    virtual int rowCount(const QModelIndex & = QModelIndex()) const override { return _data.size(); }
+    virtual QVariant data(const QModelIndex &index, int role) const override;
+
+    typedef decltype(DATA()()) type;
+    typedef DATA typestruct;
+
+private:
+    decltype(DATA()()) _data;
+};
+
+template <class DATA>
+QVariant RpgModel<DATA>::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid())
+        return QVariant();
+    else if (role == Qt::DisplayRole || role == Qt::EditRole)
+    {
+        if (index.row() == 0)
+        {
+            return QString("=== EDIT ===");
+        }
+        else
+        {
+            auto data = _data[index.row()-1];
+            return QString("%1: %2").arg(data.ID).arg(QString::fromStdString(data.name));
+        }
+    }
+    else if (role == Qt::UserRole)
+        if (index.row() > 0)
+            return _data[index.row()-1].ID;
+
+    return QVariant();
+}
+
+
+
+
+#define VARSTRUCT(STRUCTNAME, VARNAME) \
+    namespace detail { struct STRUCTNAME { decltype(Data::VARNAME) operator()() const { return Data::VARNAME; } }; } \
+    using  STRUCTNAME##RpgModel = RpgModel<detail::STRUCTNAME>;
```

Comment:
Here I define where/what the underlying data is. STRUCTNAME will be replaced by the first macro parameter (mostly the class inside the vector). This struct has an operator()() (why is explained above) that will return a reference to the RPGdata (just a return). The return value will be autodeduced by decltype.
Next line just says, that there will be (e.g.) a ActorRpgModel. the ## is stringconcat in preprocessor, and what using does is just declares something as something :D In this case you don't have to write `RpgModel<detail::Actor>` if you want an Model with an Actor, but instead just use ActorRpgModel, far more simple and hides the nasty stuff :) Hope this helps